src/nifs/protogalaxy/poly/mod.rs

```diff
@@ -322,16 +320,49 @@ pub(crate) struct PolyChallenges<F: PrimeField> {
     delta: F,
 }
 
-pub(crate) fn compute_K<F: PrimeField>(
+pub(crate) fn compute_K<F: WithSmallOrderMulGroup<3>>(
     S: &PlonkStructure<F>,
+    f_alpha: F,
     challenges: PolyChallenges<F>,
     accumulator: impl Sync + GetChallenges<F> + GetWitness<F>,
     traces: &[(impl Sync + GetChallenges<F> + GetWitness<F>)],
 ) -> Result<UnivariatePoly<F>, Error> {
-    let _poly_G = compute_G(S, challenges, accumulator, traces)?;
+    let mut g_poly = compute_G(S, challenges, accumulator, traces)?;
+    // is coeffs here, will transform to evals by fft later
+    let g_evals = g_poly.get_vector_mut();
+    let ctx = QueryIndexContext::from(S);
+    let max_degree = S
+        .gates
+        .iter()
+        .map(|poly| poly.degree(&ctx))
+        .max()
+        .unwrap_or_default();
+    let points_count = max_degree * traces.len() + 1;
+    let log_n = points_count.next_power_of_two().ilog2();
+    coset_fft(g_evals, log_n);
+
+    // eval (F(alpha)*L_0(X), Z(X)) on zeta*{1, omega,...,}
+    let l_and_z_evals = lagrange::iter_cyclic_subgroup::<F>(log_n)
+        .take(points_count)
+        .map(|pt| F::ZETA * pt)
+        .map(|pt| {
+            (
+                f_alpha * lagrange::eval_lagrange_polynomial((1 << log_n) - 1, 0, pt),
+                lagrange::eval_vanish_polynomial(1 << log_n, pt),
+            )
+        })
+        .collect::<Box<[_]>>();
+
```

Comment:
Look like this collection is don't needed, because we can zip with `g_evals` right here